File: tools/image/template_match.py
#!/usr/bin/env python3
import sys
import json
import cv2
import numpy as np

def match_template(image_path, template_path, threshold=0.7, scales="2,1.5,1,0.5"):
    img = cv2.imread(image_path, cv2.IMREAD_COLOR)
    tpl = cv2.imread(template_path, cv2.IMREAD_COLOR)
    
    if img is None or tpl is None:
        return {"matched": False, "error": "Failed to load image"}
    
    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    tpl_gray = cv2.cvtColor(tpl, cv2.COLOR_BGR2GRAY)
    
    scale_list = [float(s) for s in scales.split(",")]
    
    best_score = 0
    best_x = 0
    best_y = 0
    best_w = tpl.shape[1]
    best_h = tpl.shape[0]
    best_scale = 1.0
    
    for scale in scale_list:
        new_w = int(tpl.shape[1] * scale)
        new_h = int(tpl.shape[0] * scale)
        
        if new_w < 10 or new_h < 10:
            continue
        if new_w >= img.shape[1] or new_h >= img.shape[0]:
            continue
        
        scaled_tpl = cv2.resize(tpl_gray, (new_w, new_h))
        
        result = cv2.matchTemplate(img_gray, scaled_tpl, cv2.TM_CCOEFF_NORMED)
        min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)
        
        if max_val > best_score:
            best_score = max_val
            best_x = max_loc[0]
            best_y = max_loc[1]
            best_w = new_w
            best_h = new_h
            best_scale = scale
    
    matched = best_score >= threshold
    
    return {
        "matched": matched,
        "score": float(best_score),
        "x": int(best_x),
        "y": int(best_y),
        "centerX": int(best_x + best_w / 2),
        "centerY": int(best_y + best_h / 2),
        "width": int(best_w),
        "height": int(best_h),
        "scale": float(best_scale)
    }

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print(json.dumps({"error": "Usage: template_match.py <image> <template> [threshold] [scales]"}))
        sys.exit(1)
    
    image_path = sys.argv[1]
    template_path = sys.argv[2]
    threshold = float(sys.argv[3]) if len(sys.argv) > 3 else 0.7
    scales = sys.argv[4] if len(sys.argv) > 4 else "2,1.5,1,0.5"
    
    result = match_template(image_path, template_path, threshold, scales)
    print(json.dumps(result))

File: tools/image/template_match.mjs
import fs from "fs";
import { execFileSync } from "child_process";
import { PNG } from "pngjs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export function loadPng(imagePath) {
  if (typeof imagePath === "string") {
    const buffer = fs.readFileSync(imagePath);
    return PNG.sync.read(buffer);
  }
  return imagePath;
}

export function matchTemplate(image, template, options = {}) {
  const imgPath = typeof image === "string" ? image : null;
  const tplPath = typeof template === "string" ? template : null;
  
  if (!imgPath || !tplPath) {
    throw new Error("matchTemplate requires file paths for both image and template");
  }

  const threshold = Number.parseFloat(options.threshold || "0.7");
  const scales = options.scales || "2,1.5,1,0.5";
  
  const scriptPath = path.join(__dirname, "template_match.py");
  
  try {
    const output = execFileSync("python3", [scriptPath, imgPath, tplPath, String(threshold), scales], {
      encoding: "utf8",
      timeout: 30000,
    });
    
    const result = JSON.parse(output.trim());
    
    if (result.error) {
      throw new Error(result.error);
    }
    
    return result;
  } catch (err) {
    return {
      matched: false,
      score: 0,
      x: 0,
      y: 0,
      centerX: 0,
      centerY: 0,
      width: 0,
      height: 0,
      scale: 1,
      error: err.message,
    };
  }
}
